### PR TITLE
sds: keep warming when dynamic inserted cluster can't be extracted secret entity

### LINKED ISSUE #11120: sds: cluster not warming while certificates are being fetched; immediately marked active
When creating clusters that reference SDS certificates, the warming behavior does not seem correct. My expectation is that until the secret is sent, the cluster will be marked as "warming" until the initial_fetch_timeout, and block the rest of initialization from occuring.

What I am actually seeing is initialization *is* blocked, but there is nothing indicating the clusters are warming.

Using this config:
`docker run -v $HOME/kube/local:/config -p 15000:15000 envoyproxy/envoy-dev -c /config/envoy-sds-lds.yaml --log-format-prefix-with-location 0 --reject-unknown-dynamic-fields`

with `envoy  version: 49efb9841a58ebdc43a666f55c445911c8e4181c/1.15.0-dev/Clean/RELEASE/BoringSSL`

and config files: 

cds.yaml:
```yaml
resources:
- "@type": type.googleapis.com/envoy.config.cluster.v3.Cluster
  name: outbound_cluster_tls
  connect_timeout: 5s
  max_requests_per_connection: 1
  load_assignment:
    cluster_name: xds-grpc
    endpoints:
    - lb_endpoints:
      - endpoint:
          address:
            socket_address:
              address: 127.0.0.1
              port_value: 8080
  type: STATIC
  transport_socket:
    name: envoy.transport_sockets.tls
    typed_config:
      "@type": type.googleapis.com/envoy.api.v2.auth.UpstreamTlsContext
      common_tls_context:
        tls_certificate_sds_secret_configs:
          - name: "default"
            sds_config:
              initial_fetch_timeout: 20s
              api_config_source:
                api_type: GRPC
                grpc_services:
                  - envoy_grpc:
                      cluster_name: "sds-grpc"
                refresh_delay: 60s
        combined_validation_context:
          default_validation_context: {}
          validation_context_sds_secret_config:
            name: ROOTCA
            sds_config:
              initial_fetch_timeout: 20s
              api_config_source:
                api_type: GRPC
                grpc_services:
                - envoy_grpc:
                    cluster_name: sds-grpc    
```

envoy-sds-lds.yaml:
```yaml
admin:
  access_log_path: /dev/null
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 15000
node:
  id: id
  cluster: sdstest
dynamic_resources:
  lds_config:
    api_config_source:
      api_type: GRPC
      grpc_services:
        envoy_grpc:
          cluster_name: lds
  cds_config:
    path: /config/cds.yaml
static_resources:
  clusters:    
  - name: sds-grpc
    type: STATIC
    http2_protocol_options: {}
    connect_timeout: 5s
    lb_policy: ROUND_ROBIN
  - name: lds
    type: STATIC
    http2_protocol_options: {}
    connect_timeout: 5s
    lb_policy: ROUND_ROBIN
```

Basically what should happen here is we get a dynamic CDS cluster with SDS config. This SDS config fails, as the sds server is not setup. We have initial_fetch_timeout, so for 20s everything should be warming.

What we see instead:

* Stats are not showing warming:
```
cluster_manager.cds.init_fetch_timeout: 0
cluster_manager.cds.update_attempt: 1
cluster_manager.cds.update_failure: 0
cluster_manager.cds.update_rejected: 0
cluster_manager.cds.update_success: 1
cluster_manager.cds.update_time: 1588972075968
cluster_manager.cds.version: 17241709254077376921
cluster_manager.cluster_added: 3
cluster_manager.cluster_modified: 0
cluster_manager.cluster_removed: 0
cluster_manager.cluster_updated: 0
cluster_manager.cluster_updated_via_merge: 0
cluster_manager.update_merge_cancelled: 0
cluster_manager.update_out_of_merge_window: 0
cluster_manager.warming_clusters: 0
```

We also see `init_fetch_timeout` is `0`; this does not change after 20s

* LDS is not requested until 20s later, indicating the initial_fetch_timeout is respected. This can be seen in the logs:
(note - for simple testing I don't have a real LDS server, but we can see its not even attempted until 20s in)
```
[2020-05-08 21:07:55.967][1][info][upstream] cds: add 1 cluster(s), remove 2 cluster(s)
[2020-05-08 21:07:55.968][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:07:55.968][1][warning][config] Unable to establish new stream
[2020-05-08 21:07:55.968][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:07:55.968][1][warning][config] Unable to establish new stream
[2020-05-08 21:07:55.968][1][info][upstream] cds: add/update cluster 'outbound_cluster_tls'
[2020-05-08 21:07:55.968][1][info][main] starting main dispatch loop
[2020-05-08 21:07:56.703][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:07:56.703][1][warning][config] Unable to establish new stream
[2020-05-08 21:07:56.938][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:07:56.938][1][warning][config] Unable to establish new stream
[2020-05-08 21:07:57.135][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:07:57.135][1][warning][config] Unable to establish new stream
[2020-05-08 21:07:57.682][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:07:57.682][1][warning][config] Unable to establish new stream
[2020-05-08 21:07:58.671][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:07:58.671][1][warning][config] Unable to establish new stream
[2020-05-08 21:08:08.992][1][warning][config] StreamSecrets gRPC config stream closed: 14, no healthy upstream
[2020-05-08 21:08:08.992][1][warning][config] Unable to establish new stream
[2020-05-08 21:08:15.967][1][info][upstream] cm init: all clusters initialized
[2020-05-08 21:08:15.967][1][info][main] all clusters initialized. initializing init manager
[2020-05-08 21:08:15.967][1][warning][config] StreamListeners gRPC config stream closed: 14, no healthy upstream
```

* `dynamic_active_clusters` shows the cluster in cds.yaml. I would expect it to be "warming".

This example above is meant to simplify it, I have originally seen this with a normal deployment using ADS gRPC server (Istio) not just files.

### PR DESCRIPTION
**Pull request by Shikugawa**
Commit Message: fix #11120. This PR highly depends on #12783. Changed to keep warming if dynamic inserted clusters (when initialize doesn't finished) failed to extract TLS certificate and certificate validation context. They shouldn't be indicated as ACTIVE cluster.
Additional Description:
Risk Level: Mid
Testing: Unit
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by lizan** [comment]
@rgs1 can you try if this still crash in your case?

/wait-any

**Comment by lizan** [comment]
@mattklein123 this is ready for review.

@rgs1 friendly ping

File: source/common/runtime/runtime_features.cc [hunk 1]
```diff
@@ -112,7 +112,8 @@ constexpr const char* disabled_runtime_features[] = {
     "envoy.reloadable_features.test_feature_false",
     // gRPC Timeout header is missing (#13580)
     "envoy.reloadable_features.ext_authz_measure_timeout_on_check_created",
-
+    // The cluster which can't extract secret entity by SDS to be warming and never activate.
```
Comment (mattklein123) [inline source/common/runtime/runtime_features.cc:115]: I can't parse this. Can you describe to me what this does and I can help with some better language?

File: source/common/upstream/cluster_manager_impl.cc [hunk 2]
```diff
@@ -431,6 +431,19 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
...
+    // warming to active. In current implementation, there is no strategy to activate clusters
+    // which failed to initialize at once.
+    // TODO(shikugawa): To implement to be available by keeping warming after no-available secret
+    // entity behavior occurred. And remove
+    // `envoy.reloadable_features.cluster_keep_warming_no_secret_entity` runtime feature flag.
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.cluster_keep_warming_no_secret_entity") &&
+        !cluster.info()->transportSocketMatcher().resolve(nullptr).factory_.isReady()) {
+      ENVOY_LOG(warn, "Failed to activate {}", cluster.info()->name());
+      return;
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc:445]: How does the cluster ever activate? I think this function never gets called again?

File: source/common/upstream/cluster_manager_impl.cc [hunk 3]
```diff
@@ -431,6 +431,19 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
...
+    // If there is no secret entity, currently supports only TLS Certificate and Validation
+    // Context, when it failed to extract them via SDS, it will fail to change cluster status from
+    // warming to active. In current implementation, there is no strategy to activate clusters
+    // which failed to initialize at once.
+    // TODO(shikugawa): To implement to be available by keeping warming after no-available secret
+    // entity behavior occurred. And remove
+    // `envoy.reloadable_features.cluster_keep_warming_no_secret_entity` runtime feature flag.
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.cluster_keep_warming_no_secret_entity") &&
+        !cluster.info()->transportSocketMatcher().resolve(nullptr).factory_.isReady()) {
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc:443]: It seems strange to go through the matcher here with a null match. Shouldn't we actually be checking if *all* transport sockets are ready?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
A few comments/questions to get started. Thank you!

/wait

File: source/common/upstream/cluster_manager_impl.cc (hunk 2 above)
Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc:445]: The idea is keep the cluster warming unless it has the a valid secret entity since it can't be used in meaningful way, because the cluster will end up using NotReadySslSocket and return 503s always. That's what the TODO above is about (https://github.com/envoyproxy/envoy/issues/13952) and the long term fix https://github.com/envoyproxy/envoy/issues/13952

File: source/common/upstream/cluster_manager_impl.cc (hunk 3 above)
Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc:443]: null match is the default transport socket factory. we could check all transport factory but some of them may not be used.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:443]: IMO we don't know what is used, so we need to check if they are all ready and not go through the matcher?

File: source/common/upstream/cluster_manager_impl.cc (hunk 2 above)
Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:445]: But how can the cluster ever leave warming? It needs to get updated again by the management server?

Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc:445]: Right, until #13777 (that's the TODO in the previous comment but I pasted wrong)

**Comment by rgs1** [comment]
FYI tested this internally, no issues thus far. cc: @lizan

File: source/common/upstream/cluster_manager_impl.cc (hunk 3 above)
Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc]: Sure, done.

**Comment by lizan** [comment]
@rgs1 thanks for confirming, I think the previous commit didn't guard change correctly (it checks transport socket factory readiness regardless of runtime flag).

File: source/common/upstream/cluster_manager_impl.cc (hunk 2 above)
Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:445]: But stepping back, what does this change give us? If the cluster stays in warming and the server starts up, we will still get 503s? Or if a cluster gets stuck in warming even if the user sets a timeout, it seems like it would be extremely difficult to debug this?

What is the exact scenario we are trying to protect against? I assume the server came up and SDS is not ready during a CDS update? Because I don't think this woulds block initial server startup?

Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc:445]: Yes you still get 503s but it does affect the initialization sequence. See #11120 for more context.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:445]: Can you summarize? There is a lot of text in that issue. Or better can you just beef up the comments and then I can read that? It would be nice to have very clear comments here on how this effects initialization flow and why this is better than what we had before given that it will be warming forever.

/wait

Comment (howardjohn) [reply source/common/upstream/cluster_manager_impl.cc:445]: I think there are two related issues. The first is the cluster is marked ready immediately. This means that if SDS doesn't return the certs, we still mark everything ready. For us, we look at this readiness to determine if we should start sending traffic to the pod, so now we will start sending traffic to a ""ready"" envoy which will always fail. Ideally, if the required secrets are not there, it would be in warming, so we do not send traffic. Once the secret is returned, it could be marked ready.

The other issue, which I am not sure if it is addressed here or another PR, is that occasionally we see that Envoy fails to actually send an SDS request to the control plane, so the cluster never actually gets ready. This is exacerbated by the previous issue, because not only will we never be able to handle (tls) traffic, we will also mark ourselves as ready.

Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc:445]: @mattklein123 to be clear, this is not a completely satisfactory fix for the underlying issu. That's why it is guarded in a runtime flag and have issue for long term fix (#13952), but this solves the issue in short term, it just mitigates the not ready issue.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:445]: OK that's fine, but I don't really understand what the implications of this are and whether it's going to cause more confusing behavior that is also hard to debug. Can you add more comments and we can discuss? Do we need stats for this also? I just don't know what we are dealing with here honestly.

Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc:445]: IMO this won't cause more confusing behavior. While I agree that the current behavior is still confusing because the cluster with not ready ssl will fail when it gets traffic. The problem that this PR fix is that Envoy won't advertise itself is ready in that case.

The runtime flag is off by default and will be removed when we have long term fix. I don't think we need a stats for this given we have a warn level log and the flag is off by default.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:445]: OK can you add more comments and we can go from there? I think we are going in circles as I'm unclear on what the old/new behavior is. :)

Comment (lizan) [reply source/common/upstream/cluster_manager_impl.cc]: @mattklein123 reworded the comments.

File: source/common/upstream/cluster_manager_impl.cc [hunk 4]
```diff
@@ -431,6 +431,19 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
...
+        !cluster.info()->transportSocketMatcher().factoriesReady()) {
+      // If `envoy.reloadable_features.cluster_keep_warming_no_secret_entity` is enabled,
+      // when a cluster depends on a SDS secret but the secret entity is not ready, instead of
+      // marking it active immediately, keep it warming until the next CDS update. This let
+      // keep Envoy not advertise itself in ready state so it won't get traffic in deployments
+      // with readiness probes that checks the state.
+      // TODO(lizan): #13777/#13952 In long term we want to fix this behavior with init manager
+      // to keep clusters in warming state until Envoy get SDS response.
+      ENVOY_LOG(warn, "Failed to activate {} due to no secret entity", cluster.info()->name());
+      return;
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc:445]: I still don't understand how this blocks initialization. Unless I am missing something, the ClusterManagerInitHelper will still complete initialization with this early return. This will cause workers to start, etc., leading to 503s. If this is not the case can you update the comments?

This will block warming -> active in a subsequent CDS update, which is marginally better, but I don't think it fixes the problem of server init?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks one question.

/wait

File: source/common/upstream/cluster_manager_impl.cc (hunk 4 above)
Comment (lambdai) [reply source/common/upstream/cluster_manager_impl.cc:445]: Add the additional comment from the offline:
server init determine decide if all clusters are initialized by number of elements in `secondary_init_clusters_` and `primary_init_clusters_`
This early return doesn't skip removing that cluster from these two lists.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:445]: Right. That's what I thought.

File: source/common/upstream/cluster_manager_impl.cc (hunk 2 above)
Comment (shashankram) [reply source/common/upstream/cluster_manager_impl.cc]: > The other issue, which I am not sure if it is addressed here or another PR, is that occasionally we see that Envoy fails to actually send an SDS request to the control plane, so the cluster never actually gets ready. This is exacerbated by the previous issue, because not only will we never be able to handle (tls) traffic, we will also mark ourselves as ready.

@howardjohn I have seen such an issue as well. Was this root caused to a bug in envoy?
